Fix #4482: Minify property names ourselves in fullLink when we don't use GCC.

sjrd opened the pull request:
Alternative to #4930 that does not require predicting occurrences, nor generating the full AST twice. Based on an idea by @kyouko-taiga.

The last commit contains the change of strategy. If we go for this alternative, IMO we should squash it inside the "Minify property names" and "Compress the ancestor names".

gzm0 commented:
This is nice. Are there any downsides to this approach other than it leaks into Emitter? (FWIW, I think we can fix that, but we can look at that later). IIUC, we'll still have the same capability of compressing different partitions of names (e.g. class hierarchy that doesn't intersect) if we want to implement this (call site might have to provide more context, like the static type of the receiver, but that's probably always available).

> An alternative would be to store the counts along
with global refs in `WithGlobals`, but the overhead would then
leak pretty strongly on incremental runs that do not minify.

Huh, I didn't think about this. Do you think this will leak in performance even if we turn it off with a flag (just like dangerous v.s. non-dangerous global refs)?

sjrd commented:
> This is nice. Are there any downsides to this approach other than it leaks into Emitter? (FWIW, I think we can fix that, but we can look at that later).

I don't think so. The theoretical inelegance of having a JS AST that's not entirely immutable, but that shouldn't hold the idea.

> IIUC, we'll still have the same capability of compressing different partitions of names (e.g. class hierarchy that doesn't intersect) if we want to implement this (call site might have to provide more context, like the static type of the receiver, but that's probably always available).

Yes indeed. There should always be enough information available.

> > An alternative would be to store the counts along
> > with global refs in `WithGlobals`, but the overhead would then
> > leak pretty strongly on incremental runs that do not minify.
> 
> Huh, I didn't think about this. Do you think this will leak in performance even if we turn it off with a flag (just like dangerous v.s. non-dangerous global refs)?

I'm not sure but it might, yes.

gzm0 commented:
> I don't think so. The theoretical inelegance of having a JS AST that's not entirely immutable, but that shouldn't hold the idea.

Agreed.

> An alternative would be to store the counts along
with global refs in WithGlobals, but the overhead would then
leak pretty strongly on incremental runs that do not minify.

I have been thinking a bit more about this, and while it sounds more appealing / clean at first, I'm not sure it gives us a practical advantage as soon as post transforms come into the picture:

- It would give us re-usable JS trees, but they are ephemeral anyways once we consider post transforms.
- It does not help us / simplify control flow: In either world we have to make sure we generate all the trees before we print them.

gzm0 reviewed (changes requested):
I think we should go with this approach over #4930.

I have reviewed the minification checksizes and done a higher level review of the new approach.

I suggest we first squash before we proceed to a more detailed review.

sjrd commented:
Squashed and addressed the two simple comments. I'll look at the `PostTransformManager` comment later, but this way the rest of the PR should be reviewable.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/Printers.scala:
```diff
@@ -747,9 +747,13 @@ object Printers {
       printEscapeJS(ident.name)
 
+    protected def print(ident: DelayedIdent): Unit =
+      printEscapeJS(ident.resolveName())
```
We seem to be running into a similar issue than I had when working on the post transforms (in the version where we still had a transient tree).

We use Javascript tree printing for 2 things:
- Actual printing
- Showing a tree for debugging

In the latter case, this line is a problem: It might in case we attempt to show an unexpected tree in an exception / assertion message (because we have not generated names yet).

I'm not sure if it is worth adding options to Printers to deal with this :-/

sjrd replied:
We could probably define a subclass `JSTreeShowPrinter`, extending `JSTreePrinter` and overriding this method to show `<delayedIdent>` instead. Would that make sense?

The subclass should be better than an option on the JVM, because on the happy it's never going to be used and therefore inline caching will essentially consider it dead code.

gzm0 replied:
I think the proposed approach in the last commit looks good (modulo some details I've commented on).

I'm unsure about what the best commit structure is for it though. I feel there's a lot in the minify commit already, so maybe before, or even a separate PR (with the improved interface and a smoke test)?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:
```diff
@@ -1444,14 +1445,17 @@ private[emitter] object CoreJSLib {
+
           List(
-              MethodDef(static = false, Ident("get"), paramList(i), None, {
+              MethodDef(static = false, getName, paramList(i), None, {
```
Do the method defs here need original names?

sjrd replied:
Hum, yes, they should have one.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/SJSGen.scala:
```diff
@@ -151,29 +152,96 @@ private[emitter] final class SJSGen(
   }
 
-  def genSelect(receiver: Tree, field: irt.FieldIdent,
+  def genSelectWithOrigName(receiver: Tree, field: irt.FieldIdent,
```
What is the rationale behind adding `WithOrigName` here? IIUC the overload already defines this (and has so in the past).

Was the intention to make the `originalName` parameter optional? (which I'm not sure is a good idea, feels easier to construct an `OriginalName` at the call site where this makes sense, which IIUC would only be the CoreJSLib).

sjrd replied:
Ah, that's because I was *very* tempted to make the overload without `OriginalName` call the other one with `NoOriginalName`. However, that does *not* have the same effect. The former generates an ident that *never* has a name binding in the source map, while the latter may still generate a name binding if the produced ident does not match the source ident.

So I made it clear with the `WithOrigName` suffix that these methods do something *different*.

gzm0 replied:
I think this makes sense: It is too easy to think from just looking at `SJSGen` that we just put original names on all idents where we have them.

Maybe it makes sense to put a comment at the top explaining the `WithOrigName` suffix? Alternatively, we could call it `ForDef`, making its intent clearer (IIUC we even call `WithOrigName` if `OriginalName == NoOriginalName`, which seems like an oxymoron).

Also: Is this a refactor that is worth putting in a separate commit?

sjrd replied:
+1 for `ForDef`, and I extracted the refactoring in a separate commit.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/NameCompressor.scala:
```diff
@@ -98,11 +107,12 @@ private[emitter] object NameCompressor {
-        .thenComparing(Comparator.naturalOrder[PropertyNameEntry]()) // tie-break
+  private def allocatePropertyNames[K <: AnyRef, E <: BaseEntry with Comparable[E]](
+      entries: mutable.AnyRefMap[K, E], namesToAvoid: collection.Set[String])(
+      implicit ct: ClassTag[E]): Unit = {
```
Any reason not to use a context bound here?

gzm0 reviewed (changes requested):
Changes not concerning Emitter / BasicLinkerBackend look good. Only minor nits.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala:
```diff
@@ -39,6 +39,9 @@ final class Emitter[E >: Null <: js.Tree](
   import config._
 
+  require(!config.minify || postTransformer == PostTransformer.Identity,
+      "When using the 'minify' option, the postTransformer must be Identity.")
```
There's one significant downside with the suggested approach. It forces the `minify` code path to generate `PrintedTree`s (and their `SourceMapFragment`s), to then use them immediately and throw them away. That's a lot of work and allocations.

The current implementation directly sends their content to the module printer and source map writer.

gzm0 replied:
Yes, fair enough. So we should probably stick with this for now (my hope is that we can improve on how we do I/O, but that's for sure for later).

My hope is that we can remove the cast, but I'll comment on that elsewhere.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/Printers.scala:
```diff
@@ -842,4 +842,10 @@ object Printers {
   }
 
+  /** A printer that shows `Tree`s for debugging, not for pretty-printing. */
+  class JSTreeShowPrinter(_out: ByteArrayWriter, initIndent: Int = 0)
```
I think this makes sense. A couple of higher level points:

- Should we expose this as a method `showTree(Tree): String` only?
- I think at this point `JSTreePrinter.printTree` should probably be made protected?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/javascript/Printers.scala:
```diff
@@ -842,4 +842,10 @@ object Printers {
+  class JSTreeShowPrinter(_out: ByteArrayWriter, initIndent: Int = 0)
+      extends JSTreePrinter(_out, initIndent) {
+    override protected def print(ident: DelayedIdent): Unit =
+      print("<delayedIdent>")
```
Show the `originalName` for ease of debugging?

gzm0 commented on linker/shared/src/test/scala/org/scalajs/linker/backend/javascript/PrintersTest.scala:
```diff
@@ -159,6 +159,13 @@ class PrintersTest {
     )
   }
 
+  @Test def delayedIdentPrintVShow(): Unit = {
```
What does the `V` stand for? disjunction?

sjrd replied:
"Versus". I spelled it out completely now.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/BasicLinkerBackend.scala:
```diff
@@ -98,8 +101,15 @@ final class BasicLinkerBackend(config: LinkerBackendImpl.Config)
+            val printedTrees = (trees: List[js.Tree]).asInstanceOf[List[js.PrintedTree]]
+            for (printedTree <- printedTrees)
+              jsFileWriter.write(printedTree.jsCode)
+          }
```
Does it make sense to abstract over this with an additional class so we can avoid the cast to `List[js.PrintedTree]` with a path-dependent type)?

Like:

```scala
abstract class BodyPrinter {
  type Tree
  
  val postTransformer: PostTransformer[Tree]

  def printWithoutSourceMap(trees: List[Tree], jsFileWriter: ByteArrayWriter): Unit
  def printWithSourceMap(/* snip */)
}
```

sjrd replied:
I added a SQUASH commit that implements this suggestion. It is more verbose, but it does get rid of the cast. Let me know if it looks good to you.

gzm0 replied:
I think we should keep it yes. It's a bit more verbose, but IMO the call flow is actually easier to trace / reason about.

gzm0 reviewed (changes requested):
I've taken a look at the adjustments and replies to the comments, but I have not done a full in-detail review pass yet. (I'll do so anyways, but maybe the comments are already useful like this).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/CoreJSLib.scala:
```diff
@@ -1504,7 +1513,7 @@ private[emitter] object CoreJSLib {
         }
 
-        val clone = MethodDef(static = false, Ident(genMethodName(cloneMethodName)), Nil, None, {
+        val clone = MethodDef(static = false, genMethodIdent(cloneMethodName), Nil, None, {
```
Actually: This also would need an original name, no? (granted, that was here before, but maybe that really points towards a `genMethodIdentForDef`.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/SJSGen.scala:
```diff
@@ -32,7 +32,8 @@ import PolyfillableBuiltin._
 private[emitter] final class SJSGen(
     val jsGen: JSGen,
     val nameGen: NameGen,
```
IIUC after this commit, we shouldn't use `NameGen` directly in most places.

I initially wanted to propose to make it private here, but I'm not sure this is feasible (just yet).

But maybe it makes sense to remove the wildcard imports in `CoreJSLib`, `FunctionEmitter` and `ClassEmitter` to make the remaining call-sites more explicit? (if not, it's IMO that big of a deal, because the test suite will catch bad / partial re-namings, now that we run all of it with the minifier).

sjrd replied:
I did try that, but we still have a number of other, direct usages of `nameGen`. Notably for local names, label names, class captures, record field names. So at the moment it does not seem to be a win yet.

gzm0 reviewed (changes requested):
Full detailed review. Some minor stuff that I could have already caught earlier :shrug:

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/NameCompressor.scala:
```diff
@@ -0,0 +1,227 @@
+
+    protected def debugString: String
+
+    private object resolver extends (() => String) {
```
Have you considered making the name resolver a trait in `DelayedIdent` (in the first commit already)?

It seems all usage sites need to use `object` syntax anyways to defined it. So we might as well define two abstract methods on it to avoid forgetting to implement it at the usage site (the `toString` of a lambda will be totally useless).

sjrd replied:
It had actually occurred to me, but I did not went through.

I now introduced `DelayedIdent.Resolver` with `resolve()` and `debugString`.

gzm0 reviewed (changes requested):
The new commit structure is very nice IMO.

Just one new comment regarding the debugging interface for `DelayedIdent`.

gzm0 commented on linker/shared/src/test/scala/org/scalajs/linker/backend/javascript/PrintersTest.scala:
```diff
@@ -159,6 +163,33 @@ class PrintersTest {
+      assertEquals("x.<delayed:bar>;", tree.show)
+    }
+
+    // Even when `apply()` throws, `show` still succeeds based on `toString()`.
```
`apply()` -> `resolve()`

gzm0 reviewed (approved):
Just an outdated comment. Otherwise LGTM.

ngbinh commented:
This is exciting!
